vectorstores/cloudsql/loader/documentloader.go

```diff
@@ -0,0 +1,441 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to execute query: %w", err)
+	}
+	defer rows.Close()
+
+	fieldDescriptions := rows.FieldDescriptions()
+	columnNames := make([]string, len(fieldDescriptions))
+	valuesPrt := make([]interface{}, len(columnNames))
+
+	for i, fd := range fieldDescriptions {
+		columnNames[i] = fd.Name
+		switch fd.DataTypeOID {
+		case pgtype.TimeOID, pgtype.TimestampOID, pgtype.TimestamptzOID, pgtype.DateOID:
+			valuesPrt[i] = new(sql.NullTime)
+		case pgtype.VarcharOID, pgtype.TextOID, pgtype.JSONOID:
+			valuesPrt[i] = new(sql.NullString)
+		case pgtype.BoolOID:
+			valuesPrt[i] = new(sql.NullBool)
+		case pgtype.Float4OID, pgtype.Float8OID:
+			valuesPrt[i] = new(sql.NullFloat64)
+		case pgtype.Int2OID, pgtype.Int4OID, pgtype.Int8OID:
+			valuesPrt[i] = new(sql.NullInt64)
+		default:
+			valuesPrt[i] = new(sql.RawBytes)
+		}
+	}
+
+	for rows.Next() {
+		columnValues := make(map[string]any, len(columnNames))
+		if err := rows.Scan(valuesPrt...); err != nil {
+			return nil, fmt.Errorf("scan row failed: %v", err)
+		}
+
+		if err = rows.Scan(valuesPrt...); err != nil {
+			return nil, fmt.Errorf("failed to scan row: %w", err)
+		}
+
+		for i, name := range columnNames {
+			switch v := valuesPrt[i].(type) {
+			case *sql.NullTime:
+				if v.Valid {
+					columnValues[name] = v.Time
+				}
+			case *sql.NullString:
+				if v.Valid {
+					columnValues[name] = v.String
+				}
+			case *sql.NullBool:
+				if v.Valid {
+					columnValues[name] = v.Bool
+				}
+			case *sql.NullInt64:
+				if v.Valid {
+					columnValues[name] = v.Int64
+				}
+			case *sql.NullFloat64:
+				if v.Valid {
+					columnValues[name] = v.Float64
+				}
+			case *sql.RawBytes:
+				columnValues[name] = *v
+			default:
+				columnValues[name] = valuesPrt[i]
+			}
+		}
+
+		doc, err := l.parseDocFromRow(columnValues)
+		if err != nil {
+			return nil, err
+		}
+		documents = append(documents, doc)
+	}
+
+	if err := rows.Err(); err != nil {
+		return nil, fmt.Errorf("error during rows iteration: %w", err)
+	}
+
+	return documents, nil
+}
+
+func (l *DocumentLoader) LoadAndSplit(ctx context.Context, splitter textsplitter.TextSplitter) ([]schema.Document, error) {
+	splitteddocs := make([]schema.Document, 0)
+	if splitter == nil {
+		splitter = textsplitter.NewRecursiveCharacter()
+	}
+
+	docs, err := l.Load(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, doc := range docs {
+		contents, err := splitter.SplitText(doc.PageContent)
+		if err != nil {
+			return nil, fmt.Errorf("failed to split page content: %w", err)
+		}
+		for _, content := range contents {
+			newDoc := schema.Document{
+				PageContent: content,
+				Metadata:    doc.Metadata,
+				Score:       doc.Score,
+			}
+			splitteddocs = append(splitteddocs, newDoc)
+		}
+	}
+
+	return splitteddocs, nil
+}
+
+// NewConfig creates a new Config.
+func NewConfig(engine cloudsqlutil.PostgresEngine, options ...Option) (*Config, error) {
```

Comment:
Where is this being used? Let's follow the patterns already created for cloudsql and name this applyCloudSQLDocumentLoaderOptions